llvm/test/Transforms/InstCombine/copysign.ll

```diff
@@ -109,3 +109,18 @@ define float @fabs_mag(float %x, float %y) {
   %r = call float @llvm.copysign.f32(float %a, float %y)
   ret float %r
 }
+
+define float @copysign_conditional(i1 noundef zeroext %x, float %y, float %z) {
+; CHECK-LABEL: @copysign_conditional(
+; CHECK-NEXT:    [[CMP:%.*]] = fcmp olt float [[Y:%.*]], 0.000000e+00
+; CHECK-NEXT:    [[AND:%.*]] = and i1 [[CMP]], [[X:%.*]]
+; CHECK-NEXT:    [[SEL:%.*]] = select i1 [[AND]], float -1.000000e+00, float 1.000000e+00
+; CHECK-NEXT:    [[RES:%.*]] = tail call float @llvm.copysign.f32(float [[Z:%.*]], float [[SEL]])
+; CHECK-NEXT:    ret float [[RES]]
+;
+  %cmp = fcmp olt float %y, 0.000000e+00
+  %and = and i1 %cmp, %x
+  %sel = select i1 %and, float -1.000000e+00, float 1.000000e+00
+  %res = tail call float @llvm.copysign.f32(float %z, float %sel)
+  ret float %res
+}
```

Comment:
Need more tests covering the range of compare predicates. Also test fast math flag handling. Also should test the select with identical constants on each side

Reply:
1. I tested it for a set of predicates, the case where there’s a bit of deviation is in the oeq case, tried to handle it with specific values. On adding the below patch, this leads to a different folding (fabs and fneg) and changes results for all the predicates.
```
  /*
 oeq
 copysign(Mag, B & (A == 0.0) ? -TC : TC) --> copysign(Mag, 1) B->true.
 copysign(Mag, B & (A == 0.0) ? -TC : TC) --> copysign(Mag, -1) B->false.
  */
    Value *One;
    One = ConstantFP::get(Mag->getType(), 1.0);
   if (Pred == CmpInst::FCMP_OEQ)
     if (match(B, m_Zero()) && TC->isNegative())
       return replaceOperand(*II, 1, One);
     if (!match(B, m_Zero()) && TC->isNegative())
       One = Builder.CreateFNeg(One);
       return replaceOperand(*II, 1, One);
```
2. Added test.
3. For identical constants(cases), suppose if we have both as negative or positive, it's folded into fabs.
https://godbolt.org/z/qrfWvr465